### PR TITLE
Iox #1142 icediscovery in c

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
This PR introduces an example for service discovery in C.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1165

### CODE REVIEW

File: iceoryx_examples/icediscovery_in_c/iox_c_find_service.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+void printSearchResult(const iox_service_description_t service)
+{
+    printf(
+        "- Service: %s, Instance: %s, Event: %s\n", service.serviceString, service.instanceString, service.eventString);
+}
+
+void searchFrontCameras(const iox_service_description_t service, void* count)
+{
+    if (strcmp(service.instanceString, "FrontLeft") == 0 || strcmp(service.instanceString, "FrontRight") == 0)
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/iox_c_find_service.c:48]: There is this advice to use https://www.cplusplus.com/reference/cstring/strncmp/ to avoid errors due to missing zero termination but I am not sure we want to do this here (code gets uglier).  Up to you.

The literal is obviously zero-terminated, the strings in `service` basically are returned from our API so ...

File: iceoryx_examples/icediscovery_in_c/sleep_for.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/sleep_for.h:1]: wrong copyright?

File: iceoryx_examples/icediscovery_in_c/sleep_for.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+void sleep_for(uint32_t milliseconds)
+{
+    Sleep((uint64_t)milliseconds);
+}
+
+#else
+#include <unistd.h>
+
+void sleep_for(uint32_t milliseconds)
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/sleep_for.h:31]: Is the choice of `uint32_t` here deliberate to be able to safely multiply with 1000 and get a non-overflowing value? I think we can have an overflow in the multiplication and then things get weird (of course, this requires fairly large values > 2^32 /10³) 

According to https://man7.org/linux/man-pages/man3/usleep.3.html `useconds_t` appears to be an `unsigned int` (usually). But portability is shaky here anyway, we cannot do much except to work with microseconds in the call.

But it is only example code and not called with large values. If it were library code we would need to make it safer.

TLDR: For a lib we would need to make it safer by using `uint32_t` microseconds (and the windows call would need to perform a division by 1000). Not needed in example here.

File: iceoryx_examples/icediscovery_in_c/sleep_for.h (hunk 2 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_examples/icediscovery_in_c/sleep_for.h:1]: Nope, file copied from `icedelivery_in_c` example.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+which points to `storage`, the memory location where the service discovery is
+stored.
+
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][create service discovery handle]-->
+```c
+iox_service_discovery_storage_t storage;
+iox_service_discovery_t serviceDiscovery = iox_service_discovery_init(&storage);
+```
+
+It is included via:
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:34]: What is it? The service discovery? The handle? ... 

I would somehow start with this (to use service discovery, we need to include ...) before mpving on to service discovery creation/initialization.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][print function to be applied to search results]-->
+```c
+void printSearchResult(const iox_service_description_t service)
+{
+    printf(
+        "- Service: %s, Instance: %s, Event: %s\n", service.serviceString, service.instanceString, service.eventString);
+}
+```
+
+We can not only search for exactly matching services, but also for wildcards
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/README.md:63]: , in combination with wildcards (`NULL`) we can search also for all instances and events with the service `Radar`.

Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/README.md:63]: We can not only search for exact matching services,

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+using `NULL`:
+
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][search for all Radar services]-->
+```c
+iox_service_discovery_find_service_apply_callable(
+    serviceDiscovery, "Radar", NULL, NULL, printSearchResult, MessagingPattern_PUB_SUB);
+```
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/README.md:70]: The wildcard can be used for all strings which describe a service, e.g. service, instance and event.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                                                         NULL,
+                                                         NULL,
+                                                         searchResult,
+                                                         SEARCH_RESULT_CAPACITY,
+                                                         &missedServices,
+                                                         MessagingPattern_PUB_SUB);
+```
+
+`searchResult` is a `iox_service_description_t` array of size
+`SEARCH_RESULT_CAPACITY`. The found services are written into this array, but
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/README.md:90]: The matching services are written

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][include service discovery]-->
+```c
+#include "iceoryx_binding_c/service_discovery.h"
+```
+
+We can now call three different find service functions. Let's start with
+
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][find service and apply callable]-->
+```c
+iox_service_discovery_find_service_apply_callable(
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:45]: The terminology of callable is more of a C++ one, in C only functions and function pointers are callable and we can only pass the latter (there are no C++ functors). Still, I would keep the name as it is correct but maybe unusual to a C user.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+`Camera` services but additionally count the front camera services:
+
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][search for all front camera services]-->
+```c
+iox_service_discovery_find_service_apply_callable_with_context_data(
+    serviceDiscovery, "Camera", NULL, NULL, searchFrontDevices, &numberFrontCameras, MessagingPattern_PUB_SUB);
+```
+
+This function is quite similar to the first find service function, but we
+pass an additional argument:
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/README.md:107]: pass an additional argument `numberFrontCameras`.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][store number of front cameras]-->
+```c
+uint32_t numberFrontCameras = 0U;
+```
+
+We use this variable to store the number of front cameras and provide it as
+second parameter to the function which is applied to all found services:
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/README.md:115]: and provide a pointer to it as second argument `count` to the function `searchFrontDevices` which is applied

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                                                         NULL,
+                                                         searchResult,
+                                                         SEARCH_RESULT_CAPACITY,
+                                                         &missedServices,
+                                                         MessagingPattern_PUB_SUB);
+```
+
+`searchResult` is a `iox_service_description_t` array of size
+`SEARCH_RESULT_CAPACITY`. The found services are written into this array, but
+at maximum `SEARCH_RESULT_CAPACITY`. If the number of found services exceeds
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:91]:
```suggestion
up to a maximum of `SEARCH_RESULT_CAPACITY`. If the number of found services exceeds
```

File: iceoryx_examples/icediscovery_in_c/README.md (hunk 7 above)
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:90]:
```suggestion
`SEARCH_RESULT_CAPACITY`. The found services are written into this array
```

File: iceoryx_examples/icediscovery_in_c/iox_c_find_service.c [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+            printSearchResult(searchResult[i]);
+        }
+
+        //! [search for all front camera services]
+        iox_service_discovery_find_service_apply_callable_with_context_data(
+            serviceDiscovery, "Camera", NULL, NULL, searchFrontDevices, &numberFrontCameras, MessagingPattern_PUB_SUB);
+        //! [search for all front camera services]
+        printf("\nFound %u front cameras\n", numberFrontCameras);
+
+        sleep_for(1000);
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/iox_c_find_service.c:128]: Could you please remove the magic number here:
```c
const uint32_t WAIT_TIME_IN_MS = 1000;
sleep_for(WAIT_TIME_IN_MS);
```

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+the array's capacity, the number of not stored services is written to
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:92]:
```suggestion
the array's capacity, the number of services that could not be stored is written to
```

File: iceoryx_examples/icediscovery_in_c/README.md (hunk 10 above)
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:115]: Maybe it is more important that it is an output parameter to return the count. Do we want to add this information or is this clear from usage? To me it is but I may be biased.

File: iceoryx_examples/icediscovery_in_c/iox_c_find_service.c [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+{
+    printf(
+        "- Service: %s, Instance: %s, Event: %s\n", service.serviceString, service.instanceString, service.eventString);
+}
+//! [print function to be applied to search results]
+
+//! [search function for all front devices]
+void searchFrontDevices(const iox_service_description_t service, void* count)
+{
+    if (strncmp(service.instanceString, "FrontLeft", IOX_CONFIG_SERVICE_STRING_SIZE - 1U) == 0
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/iox_c_find_service.c:53]: This `-1U` here on user side is unfortunate and error-prone. Do we want to define the string size *without*  the zero terminator included and compensate with a +1 in the tests instead where needed?

Or even have two constants, i.e. `IOX_CONFIG_SERVICE_STRING_SIZE_WITH_ZERO_TERMINATON` in addition.

From https://www.cplusplus.com/reference/cstring/strncmp/ I am not sure we really need to subtract 1 here though for correctness. It will not exceed the buffer either way. In this case we would only drop the `-1U`.

**Review by MatthiasKillat — COMMENTED** [review]
Looks good, some final minor changes should be considered.

File: iceoryx_examples/icediscovery_in_c/sleep_for.h (hunk 3 above)
Comment (MatthiasKillat) [reply iceoryx_examples/icediscovery_in_c/sleep_for.h:31]: @FerdinandSpitzschnueffler I think this is of no concern as we use it in the example and can be resolved.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+To be able to search for services, we need to include:
+
+<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][include service discovery]-->
+```c
+#include "iceoryx_binding_c/service_discovery.h"
+```
+
+To get a handle to the service discovery we call `iox_service_discovery_init`.
+The service discovery requires memory of the amount of
+`sizeof(iox_service_discovery_storage_t)` to be initialized.
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:33]: Ok, but this is hidden from the user and an implementation detail. From my point of view the important information is that the user must provide an instance of `iox_service_discovery_storage_t` to an initialization function. The size of objects of this class is not relevant to the user and taking care of by the struct itself.

Technically the user can pass any buffer of sufficient size (requires a cast), but it is not intended to be used that way...

I would hence not mention the detail wit `sizeof`, but not a must change ...

**Review by MatthiasKillat — DISMISSED** [review]

File: iceoryx_examples/icediscovery_in_c/README.md (hunk 15 above)
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:33]: Technically the only information relevant to the user is that she has to pass a struct of type `iox_service_discovery_storage_t` which will be of the correct size (by construction/API). 

Hence one could leave out the information with `sizeof`. It may give people ideas to use it with raw buffers of that size (which will work if correctly aligned but is not encouraged).

Not a must change but should probably be omitted.

File: iceoryx_binding_c/include/iceoryx_binding_c/service_description.h [hunk 16]
```diff
@@ -17,13 +17,15 @@
...
+#include "iceoryx_binding_c/config.h"
+
 #include <stdint.h>
 
 typedef struct
 {
-    char serviceString[100U];
-    char instanceString[100U];
-    char eventString[100U];
+    char serviceString[IOX_CONFIG_SERVICE_STRING_SIZE + 1U]; // +1U for \0 terminator
```
Comment (elfenpiff) [inline iceoryx_binding_c/include/iceoryx_binding_c/service_description.h:26]: Please do not do this. `IOX_CONFIG_SERVICE_STRING_SIZE` should contain the actual capacity of the string including the null terminator.
Otherwise we have a weird behavior when for instance we have `sizeof` calculations etc.

File: iceoryx_binding_c/include/iceoryx_binding_c/config.h [hunk 17]
```diff
@@ -97,4 +97,7 @@ uint32_t iox_cfg_max_runtime_name_length();
 /// @brief the maximum size of a node name string + \0 terminator
 #define IOX_CONFIG_NODE_NAME_SIZE 101
 
+/// @brief the maximum size of a service description string identifier
+#define IOX_CONFIG_SERVICE_STRING_SIZE 100
```
Comment (elfenpiff) [inline iceoryx_binding_c/include/iceoryx_binding_c/config.h:101]: Please change it to 101

File: iceoryx_binding_c/test/moduletests/test_service_description.cpp [hunk 18]
```diff
@@ -32,13 +32,14 @@ using namespace iox::capro;
 
 TEST(iox_service_description_test, StringSizesAreCorrect)
 {
+    constexpr uint64_t ZERO_TERMINATOR_SIZE = 1U;
     ::testing::Test::RecordProperty("TEST_ID", "f32a6d19-9ff7-4913-a1d8-39d46267b114");
     EXPECT_THAT(sizeof(decltype(std::declval<iox_service_description_t>().serviceString)),
-                Eq(iox::capro::IdString_t().capacity()));
+                Eq(iox::capro::IdString_t().capacity() + ZERO_TERMINATOR_SIZE));
```
Comment (elfenpiff) [inline iceoryx_binding_c/test/moduletests/test_service_description.cpp:38]: The null terminator is always included in string literals, see
```c
const char blubb[] = "EvilHuhn";
printf("%lu \n", sizeof(blubb) / sizeof(char) ); // = 9, since the zero terminator is always part of the array.
```

File: iceoryx_examples/icediscovery_in_c/iox_c_find_service.c [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+{
+    printf(
+        "- Service: %s, Instance: %s, Event: %s\n", service.serviceString, service.instanceString, service.eventString);
+}
+//! [print function to be applied to search results]
+
+//! [search function for all front devices]
+void searchFrontDevices(const iox_service_description_t service, void* count)
+{
+    if (strncmp("FrontLeft", service.instanceString, IOX_CONFIG_SERVICE_STRING_SIZE) == 0
```
Comment (elfenpiff) [inline iceoryx_examples/icediscovery_in_c/iox_c_find_service.c:53]: If clang-tidy spits out a warning here, please ignore it, see this little thing for instance:
```c
 #include <stdio.h>
 #include <string.h>
 #define ID_NAME 6
 int main() {
     const char service[ID_NAME] = "12453";
     return strncmp("12", service, 15); // clang-tidy warning
     return strncmp("12", service, 15 - 0); // NO clang-tidy warning
     return strncmp("12", service, 15 - 1230); // NO clang-tidy warning
 }

```

It may makes sense to add `-bugprone-not-null-terminated-result` as exclusion to our root `.clang-tidy` file

File: iceoryx_binding_c/include/iceoryx_binding_c/service_description.h (hunk 16 above)
Comment (MatthiasKillat) [reply iceoryx_binding_c/include/iceoryx_binding_c/service_description.h:26]: I would argue the string size is n and the buffer size is n +1.

File: iceoryx_binding_c/test/moduletests/test_service_description.cpp (hunk 18 above)
Comment (MatthiasKillat) [reply iceoryx_binding_c/test/moduletests/test_service_description.cpp:38]: For literals sure but `IdString_t::capacity` returns a capacity which excludes the null terminator.

File: iceoryx_binding_c/include/iceoryx_binding_c/config.h (hunk 17 above)
Comment (MatthiasKillat) [reply iceoryx_binding_c/include/iceoryx_binding_c/config.h:101]: This is not very intuitive as it is the buffer size.

File: iceoryx_examples/icediscovery_in_c/README.md [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+                                                         &missedServices,
+                                                         MessagingPattern_PUB_SUB);
+```
+
+`searchResult` is a `iox_service_description_t` array of size
+`SEARCH_RESULT_CAPACITY`. The matching services are written into this array, up
+to a maximum of `SEARCH_RESULT_CAPACITY`. If the number of found services
+exceeds the array's capacity, the number of services that could not be stored is
+written to `missedServices`. The number of stored services is returned. Since
+the `cameraPublishers` periodically offer/stop their services, you should see
```
Comment (MatthiasKillat) [inline iceoryx_examples/icediscovery_in_c/README.md:94]:
```suggestion
the `cameraPublishers` periodically offer/stop their services, we should see
```

**Review by MatthiasKillat — APPROVED** [review]

File: iceoryx_examples/icediscovery_in_c/README.md (hunk 10 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_examples/icediscovery_in_c/README.md]: I find this clear from usage. Additionally we state `We use this variable to store the number...`

File: iceoryx_examples/icediscovery_in_c/iox_c_find_service.c (hunk 19 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_examples/icediscovery_in_c/iox_c_find_service.c]: I added the exclusion

**Review by elfenpiff — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]
